### PR TITLE
Searcher: return detected language

### PR DESCRIPTION
**Pull request by jtibshirani**
For searches that hit Zoekt or the symbols backend, we return the file's
detected language through `result.File.PreciseLanguage`. Now, searcher will
also return the detected language whenever it's available. This value is used
in the search results filters panel, so it's important to be as accurate as
possible.

The precise language is only available when these criteria are met:
* The `search-content-based-lang-detection` feature is enabled -- we hope to
enable this by default soon
* The search contains a language filter, so we've already run language
detection

The PR also adds an optimization to run the expensive language detection
only after checking the search pattern.

Relates to https://github.com/sourcegraph/sourcegraph/issues/60341

## Test plan

Updated unit tests, manual testing

### CODE REVIEW

**Comment by jtibshirani** [comment]
This is a big improvement, but still leaves some cases where we don't have the precise language. Here's an example where the filters panel can still be inaccurate (this is a C++ only repo):
Screenshot 2024-02-19 at 1 58 15 PM
We have a few options:
1. Always run language detection on every match, even if there are no language filters. 
2. Accept the inconsistency.
3. Don't display language filters in the filters panel unless we have `PreciseLanguage` results. In practice, this means indexed and symbol searches are fine, but some unindexed searches are missing the lang filters in the panel.

I don't like option (1) because it would significantly slow down unindexed search in some common cases. Even for a `type:path` search, we would always have to load file contents, and always run go-enry. My preferred approach is (3). What do you think?

**Comment by camdencheek** [comment]
> it would significantly slow down unindexed search

Have you measured this? I was also corrected by @varungandhi-src on this, but the language detection only actually calls the "get the file contents" function if there is ambiguity, which there isn't for most file extensions. So I imagine this would impact the speed of C/C++ codebases, but not, for example, Go.

File: cmd/searcher/internal/search/search_regex.go [hunk 1]
```diff
@@ -168,7 +161,12 @@ func regexSearch(
 				}
 
 				if match {
-					sender.Send(fm)
+					// Apply language filters and send result
```
Comment (jtibshirani) [inline cmd/searcher/internal/search/search_regex.go:164]: This is just an optimization I snuck in that I noticed during benchmarks. It's not central to the change.

**Comment by jtibshirani** [comment]
I experimented with this patch to always run lang detection, even when there are no lang filters:

<details>

```diff
diff --git a/cmd/searcher/internal/search/langmatch.go b/cmd/searcher/internal/search/langmatch.go
index 33ef654049f..691d2c6646f 100644
--- a/cmd/searcher/internal/search/langmatch.go
+++ b/cmd/searcher/internal/search/langmatch.go
@@ -7,31 +7,19 @@ import (
 	"github.com/sourcegraph/sourcegraph/lib/codeintel/languages"
 )
 
-// langMatcher checks whether a file matches the 'include' and 'exclude
-// language filters
-type langMatcher interface {
-	// Matches checks whether a file's language matches. It accepts a callback
-	// to fetch the content to avoid loading content when it's not needed. It
-	// returns whether the file matches, plus the detected language when available.
-	Matches(path string, getContent func() ([]byte, error)) (bool, string)
-}
-
-type allLangMatcher struct{}
-
-func (am *allLangMatcher) Matches(_ string, _ func() ([]byte, error)) (bool, string) {
-	return true, ""
-}
-
-// enryLangMatcher uses go-ery to check whether files match the language
-// filters. IncludeLangs and ExcludeLangs are expected to be valid languages,
-// like from the output of enry.GetLanguagesByAlias.
-type enryLangMatcher struct {
+// langMatcher checks whether a file matches the 'include' and 'exclude language filters.
+// It uses go-enry to check whether files match. IncludeLangs and ExcludeLangs are expected
+// to be valid languages, like from the output of enry.GetLanguagesByAlias.
+type langMatcher struct {
 	IncludeLangs []string
 	ExcludeLangs []string
 	OnlyExcludes bool
 }
 
-func (em *enryLangMatcher) Matches(path string, getContent func() ([]byte, error)) (bool, string) {
+// Matches checks whether a file's language matches. It accepts a callback
+// to fetch the content to avoid loading content when it's not needed. It
+// returns whether the file matches, plus the detected language when available.
+func (lm *langMatcher) Matches(path string, getContent func() ([]byte, error)) (bool, string) {
 	// We use Sourcegraph's wrapper around enry because it supports lazily fetching
 	// content and contains some optimizations for ambiguous extensions.
 	langs, err := languages.GetLanguages(path, getContent)
@@ -43,26 +31,21 @@ func (em *enryLangMatcher) Matches(path string, getContent func() ([]byte, error
 
 	// It's fine if file has no detected language, as long as there are no include filters
 	if len(langs) == 0 {
-		return em.OnlyExcludes, ""
+		return lm.OnlyExcludes, ""
 	}
 
 	// Choose the most likely language
 	lang := langs[0]
-	for _, includeLang := range em.IncludeLangs {
+	for _, includeLang := range lm.IncludeLangs {
 		if lang != includeLang {
 			return false, ""
 		}
 	}
-	return !slices.Contains(em.ExcludeLangs, lang), lang
+	return !slices.Contains(lm.ExcludeLangs, lang), lang
 }
 
 func toLangMatcher(p *protocol.PatternInfo) langMatcher {
-	// If there are no language filters, avoid all path and content checks.
-	if len(p.IncludeLangs)+len(p.ExcludeLangs) == 0 {
-		return &allLangMatcher{}
-	}
-
-	return &enryLangMatcher{
+	return langMatcher{
 		IncludeLangs: p.IncludeLangs,
 		ExcludeLangs: p.ExcludeLangs,
 		OnlyExcludes: len(p.IncludeLangs) == 0,
diff --git a/cmd/searcher/internal/search/search_regex_test.go b/cmd/searcher/internal/search/search_regex_test.go
index 0fa4c4a2c8e..3298d1b6851 100644
--- a/cmd/searcher/internal/search/search_regex_test.go
+++ b/cmd/searcher/internal/search/search_regex_test.go
@@ -512,7 +512,7 @@ func TestRegexSearch(t *testing.T) {
 				},
 				zf: file,
 			},
-			wantFm: []protocol.FileMatch{{Path: "a.go"}},
+			wantFm: []protocol.FileMatch{{Path: "a.go", Language: "Go"}},
 		},
 		{
 			name: "'and' query with matches",
@@ -531,7 +531,8 @@ func TestRegexSearch(t *testing.T) {
 				zf: file,
 			},
 			wantFm: []protocol.FileMatch{{
-				Path: "a.go",
+				Path:     "a.go",
+				Language: "Go",
 				ChunkMatches: []protocol.ChunkMatch{{
 					Content:      "aaaaa11111",
 					ContentStart: protocol.Location{0, 0, 0},
@@ -573,7 +574,7 @@ func TestRegexSearch(t *testing.T) {
 				},
 				zf: file,
 			},
-			wantFm: []protocol.FileMatch{{Path: "a.go"}},
+			wantFm: []protocol.FileMatch{{Path: "a.go", Language: "Go"}},
 		},
 		{
 			name: "'or' query with matches",
@@ -592,7 +593,8 @@ func TestRegexSearch(t *testing.T) {
 				zf: file,
 			},
 			wantFm: []protocol.FileMatch{{
-				Path: "a.go",
+				Path:     "a.go",
+				Language: "Go",
 				ChunkMatches: []protocol.ChunkMatch{{
 					Content:      "aaaaa11111",
 					ContentStart: protocol.Location{0, 0, 0},
@@ -643,7 +645,7 @@ func TestRegexSearch(t *testing.T) {
 				},
 				zf: file,
 			},
-			wantFm: []protocol.FileMatch{{Path: "a.go"}},
+			wantFm: []protocol.FileMatch{{Path: "a.go", Language: "Go"}},
 		},
 	}
 
@@ -671,6 +673,7 @@ func TestLangFilters(t *testing.T) {
 		"a.go":      "aaaaa11111",
 		"README.md": "important info on go",
 		"file.m":    "[x,y,z] = sphere; \nr = 2;\nsurf(x*r,y*r,z*r)\naxis equal",
+		"unknown":   "1234567",
 	})
 	file, _ := mockZipFile(zipData)
 
@@ -787,6 +790,20 @@ func TestLangFilters(t *testing.T) {
 			},
 			wantFm: nil,
 		},
+		{
+			name: "no lang filters",
+			args: args{
+				ctx: context.Background(),
+				p: &protocol.PatternInfo{
+					Query:                 &protocol.PatternNode{Value: ""},
+					IncludePaths:          []string{"unknown"},
+					PatternMatchesPath:    false,
+					PatternMatchesContent: true,
+				},
+				zf: file,
+			},
+			wantFm: []protocol.FileMatch{{Path: "unknown"}},
+		},
 	}
 
 	for _, tt := range tests {
```
</details>

Here are the new benchmark results. There is a significant regression, especially for "SearchRegex_large_empty_pattern", which makes sense as we're now running language detection over a large result set. We're now allocating > 100MB per op, and the test repo `golang/go` is relatively small with ~6500 files.

<details>

```
goos: darwin
goarch: arm64
pkg: github.com/sourcegraph/sourcegraph/cmd/searcher/internal/search
                                                │   old.txt    │                 new.txt                 │
                                                │    sec/op    │     sec/op      vs base                 │
SearchRegex_large_fixed-10                        5.373m ± ∞ ¹     5.452m ± ∞ ¹     +1.46% (p=0.008 n=5)
SearchRegex_rare_fixed-10                         7.138m ± ∞ ¹     7.456m ± ∞ ¹     +4.46% (p=0.008 n=5)
SearchRegex_large_fixed_casesensitive-10          2.822m ± ∞ ¹     2.829m ± ∞ ¹          ~ (p=0.310 n=5)
SearchRegex_large_empty_pattern-10                5.250m ± ∞ ¹   149.206m ± ∞ ¹  +2741.81% (p=0.008 n=5)
SearchRegex_large_lang_filter_common-10           4.728m ± ∞ ¹     4.281m ± ∞ ¹     -9.44% (p=0.008 n=5)
SearchRegex_large_lang_filter_uncommon-10         1.300m ± ∞ ¹     1.070m ± ∞ ¹    -17.63% (p=0.008 n=5)
SearchRegex_large_re_dotstar-10                   238.0m ± ∞ ¹     359.1m ± ∞ ¹    +50.87% (p=0.008 n=5)
SearchRegex_large_re_common-10                    5.778m ± ∞ ¹    10.842m ± ∞ ¹    +87.64% (p=0.008 n=5)
SearchRegex_large_re_anchor-10                    57.43m ± ∞ ¹     57.01m ± ∞ ¹          ~ (p=0.151 n=5)
SearchRegex_large_capture_group-10                138.4m ± ∞ ¹     165.7m ± ∞ ¹    +19.73% (p=0.008 n=5)
SearchRegex_large_path/path_only-10               765.2µ ± ∞ ¹     739.1µ ± ∞ ¹     -3.40% (p=0.008 n=5)
SearchRegex_large_path/content_only-10            6.987m ± ∞ ¹     7.029m ± ∞ ¹          ~ (p=0.310 n=5)
SearchRegex_large_path/both_path_and_content-10   7.019m ± ∞ ¹     6.969m ± ∞ ¹          ~ (p=0.151 n=5)
SearchRegex_small_fixed-10                        141.6µ ± ∞ ¹     144.2µ ± ∞ ¹     +1.86% (p=0.008 n=5)
SearchRegex_small_fixed_casesensitive-10          99.45µ ± ∞ ¹    100.05µ ± ∞ ¹          ~ (p=0.056 n=5)
SearchRegex_small_re_dotstar-10                   2.262m ± ∞ ¹     2.292m ± ∞ ¹     +1.31% (p=0.008 n=5)
SearchRegex_small_re_common-10                    103.1µ ± ∞ ¹     117.0µ ± ∞ ¹    +13.48% (p=0.008 n=5)
SearchRegex_small_re_anchor-10                    724.0µ ± ∞ ¹     740.7µ ± ∞ ¹          ~ (p=0.095 n=5)
SearchRegex_small_capture_group-10                1.574m ± ∞ ¹     1.597m ± ∞ ¹     +1.46% (p=0.032 n=5)
geomean                                           3.251m           4.119m          +26.68%
¹ need >= 6 samples for confidence interval at level 0.95

                                                │    old.txt    │                new.txt                 │
                                                │     B/op      │      B/op       vs base                │
SearchRegex_large_fixed-10                        14.97Mi ± ∞ ¹    14.97Mi ± ∞ ¹         ~ (p=0.421 n=5)
SearchRegex_rare_fixed-10                         14.97Mi ± ∞ ¹    14.97Mi ± ∞ ¹         ~ (p=0.095 n=5)
SearchRegex_large_fixed_casesensitive-10          14.83Ki ± ∞ ¹    15.35Ki ± ∞ ¹    +3.55% (p=0.008 n=5)
SearchRegex_large_empty_pattern-10                16.00Mi ± ∞ ¹   102.70Mi ± ∞ ¹  +541.85% (p=0.008 n=5)
SearchRegex_large_lang_filter_common-10           14.93Mi ± ∞ ¹    14.93Mi ± ∞ ¹         ~ (p=0.548 n=5)
SearchRegex_large_lang_filter_uncommon-10         14.93Mi ± ∞ ¹    14.65Mi ± ∞ ¹    -1.89% (p=0.008 n=5)
SearchRegex_large_re_dotstar-10                   599.8Mi ± ∞ ¹    687.8Mi ± ∞ ¹   +14.66% (p=0.008 n=5)
SearchRegex_large_re_common-10                    7.150Mi ± ∞ ¹   11.647Mi ± ∞ ¹   +62.88% (p=0.008 n=5)
SearchRegex_large_re_anchor-10                    4.691Mi ± ∞ ¹    4.826Mi ± ∞ ¹    +2.88% (p=0.008 n=5)
SearchRegex_large_capture_group-10                3.299Mi ± ∞ ¹   18.142Mi ± ∞ ¹  +449.93% (p=0.008 n=5)
SearchRegex_large_path/path_only-10               8.571Ki ± ∞ ¹    8.619Ki ± ∞ ¹    +0.56% (p=0.008 n=5)
SearchRegex_large_path/content_only-10            64.70Ki ± ∞ ¹    65.75Ki ± ∞ ¹    +1.63% (p=0.008 n=5)
SearchRegex_large_path/both_path_and_content-10   63.50Ki ± ∞ ¹    66.55Ki ± ∞ ¹    +4.81% (p=0.016 n=5)
SearchRegex_small_fixed-10                        320.5Ki ± ∞ ¹    342.8Ki ± ∞ ¹    +6.96% (p=0.008 n=5)
SearchRegex_small_fixed_casesensitive-10          10.56Ki ± ∞ ¹    10.54Ki ± ∞ ¹    -0.28% (p=0.016 n=5)
SearchRegex_small_re_dotstar-10                   3.934Mi ± ∞ ¹    3.978Mi ± ∞ ¹    +1.12% (p=0.008 n=5)
SearchRegex_small_re_common-10                    48.43Ki ± ∞ ¹    50.30Ki ± ∞ ¹    +3.87% (p=0.008 n=5)
SearchRegex_small_re_anchor-10                    26.22Ki ± ∞ ¹    27.08Ki ± ∞ ¹    +3.30% (p=0.008 n=5)
SearchRegex_small_capture_group-10                24.75Ki ± ∞ ¹    25.12Ki ± ∞ ¹         ~ (p=0.056 n=5)
geomean                                           800.8Ki         1012.8Ki         +26.47%
¹ need >= 6 samples for confidence interval at level 0.95

                                                │   old.txt    │                  new.txt                  │
                                                │  allocs/op   │   allocs/op     vs base                   │
SearchRegex_large_fixed-10                         260.0 ± ∞ ¹      260.0 ± ∞ ¹            ~ (p=0.444 n=5)
SearchRegex_rare_fixed-10                          677.0 ± ∞ ¹      679.0 ± ∞ ¹       +0.30% (p=0.048 n=5)
SearchRegex_large_fixed_casesensitive-10           211.0 ± ∞ ¹      212.0 ± ∞ ¹       +0.47% (p=0.008 n=5)
SearchRegex_large_empty_pattern-10                 121.0 ± ∞ ¹   519446.0 ± ∞ ¹  +429194.21% (p=0.008 n=5)
SearchRegex_large_lang_filter_common-10            317.0 ± ∞ ¹      285.0 ± ∞ ¹      -10.09% (p=0.008 n=5)
SearchRegex_large_lang_filter_uncommon-10          351.0 ± ∞ ¹      197.0 ± ∞ ¹      -43.87% (p=0.008 n=5)
SearchRegex_large_re_dotstar-10                   6.099M ± ∞ ¹     6.619M ± ∞ ¹       +8.53% (p=0.008 n=5)
SearchRegex_large_re_common-10                    68.79k ± ∞ ¹    100.34k ± ∞ ¹      +45.86% (p=0.008 n=5)
SearchRegex_large_re_anchor-10                    50.16k ± ∞ ¹     50.18k ± ∞ ¹       +0.04% (p=0.008 n=5)
SearchRegex_large_capture_group-10                25.29k ± ∞ ¹    108.68k ± ∞ ¹     +329.79% (p=0.008 n=5)
SearchRegex_large_path/path_only-10                129.0 ± ∞ ¹      130.0 ± ∞ ¹       +0.78% (p=0.032 n=5)
SearchRegex_large_path/content_only-10             669.0 ± ∞ ¹      671.0 ± ∞ ¹       +0.30% (p=0.008 n=5)
SearchRegex_large_path/both_path_and_content-10    647.0 ± ∞ ¹      649.0 ± ∞ ¹       +0.31% (p=0.008 n=5)
SearchRegex_small_fixed-10                         146.0 ± ∞ ¹      147.0 ± ∞ ¹       +0.68% (p=0.008 n=5)
SearchRegex_small_fixed_casesensitive-10           139.0 ± ∞ ¹      140.0 ± ∞ ¹       +0.72% (p=0.008 n=5)
SearchRegex_small_re_dotstar-10                   40.16k ± ∞ ¹     40.17k ± ∞ ¹       +0.03% (p=0.008 n=5)
SearchRegex_small_re_common-10                     547.0 ± ∞ ¹      549.0 ± ∞ ¹       +0.37% (p=0.008 n=5)
SearchRegex_small_re_anchor-10                     351.0 ± ∞ ¹      352.0 ± ∞ ¹       +0.28% (p=0.008 n=5)
SearchRegex_small_capture_group-10                 332.0 ± ∞ ¹      333.0 ± ∞ ¹       +0.30% (p=0.008 n=5)
geomean                                           1.419k           2.357k            +66.12%
¹ need >= 6 samples for confidence interval at level 0.95
```
</details>

I checked and the allocations are mostly due to lang detection (even the first two regexp functions are due to `go-enry`):
```
Showing top 10 nodes out of 48
      flat  flat%   sum%        cum   cum%
  878.71MB 21.15% 21.15%   928.81MB 22.36%  regexp.(*Regexp).replaceAll
  717.35MB 17.27% 38.42%   717.35MB 17.27%  regexp.(*Regexp).FindAll.func1
  709.35MB 17.08% 55.50%  3324.38MB 80.03%  github.com/go-enry/go-enry/v2/internal/tokenizer.Tokenize
  596.46MB 14.36% 69.86%   596.46MB 14.36%  github.com/sourcegraph/sourcegraph/cmd/searcher/internal/search.(*fileLoader).load
  250.07MB  6.02% 75.88%   250.07MB  6.02%  github.com/go-enry/go-enry/v2/internal/tokenizer.toString (inline)
  223.72MB  5.39% 81.27%   525.79MB 12.66%  github.com/go-enry/go-enry/v2/internal/tokenizer.extractRemainders
```

I'm inclined not to enable language detection by default for this PR. For now I'd document the inconsistency and we can revisit our options when deciding whether to enable the feature flag by default. Does this work for you?

**Comment by camdencheek** [comment]
Wow, I'm surprised that language detection is that expensive. In particular, I would not have expected the library to be allocating, like, at all. Or calling `regexp.ReplaceAll()`. Thank you for doing the benchmarks! 

As a followup, it could be valuable to figure out which of the enry match strategies are contributing the most to the perf hit and whether they are actually providing much value, but in the mean time, I think it's probably fine to only do precise language detection when a lang filter is present.

My biggest concern is the filters panel: if we are only doing precise detection when a lang filter is present, won't this contribute to _even more_ of a mismatch between what we show in the filters panel and what is actually returned when we search? We will use the file name to determine which filters to show, but if you click that filter, when we actually run the search, we will use precise language detection which might determine the file is actually for a different language and not return any results.

**Comment by jtibshirani** [comment]
> My biggest concern is the filters panel: if we are only doing precise detection when a lang filter is present, won't this contribute to even more of a mismatch between what we show in the filters panel and what is actually returned when we search?

Indeed, this is the main inconsistency here. To clarify, it only applies when the search is (1) unindexed, and (2) there are no lang filters. In addition, when hybrid search is used, we use the precise language directly from Zoekt, so in practice most files will still have the precise lang.

You could argue that we currently have a pretty big inconsistency (with `search-content-based-lang-detection: false`), since the filters panel uses `PreciseLanguage`, but search filters use file extensions. Here's [an example of an indexed search](https://sourcegraph.com/search?q=context:global+repo:grf-labs/grf+lang:cpp&patternType=keyword&sm=0) where we apply a `lang:cpp` filter, but the filters panel shows both "C" and "skipped" langs.

**Review by camdencheek — APPROVED** [review]

**Pull request merged**